src/Microsoft.DotNet.ProjectModel/Workspace/WorkspaceContext.cs

```diff
@@ -0,0 +1,328 @@
+        /// 
+        /// There must be either a global.json or project.json at under the given path. Otherwise
+        /// null is returned.
+        /// </summary>
+        public static WorkspaceContext CreateFrom(string projectPath, string configuration)
+        {
+            var projectPaths = PathResolve(projectPath);
+            if (projectPaths == null || !projectPaths.Any())
+            {
+                return null;
+            }
+
+            var context = new WorkspaceContext(projectPaths, configuration);
+            return context;
+        }
+
+        public static WorkspaceContext CreateFrom(string projectPath)
+        {
+            return CreateFrom(projectPath, "Debug");
+        }
+
+        private void Initialize(List<string> projectPaths)
+        {
+            _projects.Clear();
+            foreach (var projectDirectory in projectPaths)
+            {
+                var project = GetProject(projectDirectory);
+
+                if (project == null)
+                {
+                    continue;
+                }
+
+                _projects.Add(project.ProjectFilePath);
+
+                foreach (var framework in project.GetTargetFrameworks())
+                {
+                    var projectReferences = GetProjectReferences(project, framework.FrameworkName);
+                    foreach (var eachReference in projectReferences)
+                    {
+                        var referencedProject = GetProject(eachReference.Path);
+                        if (referencedProject == null)
+                        {
+                            continue;
+                        }
+
+                        _projects.Add(referencedProject.ProjectFilePath);
+                    }
+                }
+            }
+        }
+
+        public IEnumerable<string> Projects => _projects;
+
+        public Project GetProject(string projectDirectory)
+        {
+            return _projectsCache.AddOrUpdate(
+                projectDirectory,
+                key => AddProjectEntry(key, null),
+                (key, oldEntry) => AddProjectEntry(key, oldEntry)).Model;
+        }
+
+        public LockFile GetLockFile(string projectDirectory)
+        {
+            return _lockFileCache.AddOrUpdate(
+                projectDirectory,
+                key => AddLockFileEntry(key, null),
+                (key, oldEntry) => AddLockFileEntry(key, oldEntry)).Model;
+        }
+
+        public ProjectContext GetProjectContext(Project project, NuGetFramework framework)
+        {
+            return _projectContextsCache.AddOrUpdate(
+                Tuple.Create(project.ProjectDirectory, framework),
+                key => AddProjectContextEntry(key.Item1, key.Item2, null),
+                (key, oldEntry) => AddProjectContextEntry(key.Item1, key.Item2, oldEntry)).ProjectContext;
+        }
+
+        public IEnumerable<DependencyDescription> GetDependencies(Project project, NuGetFramework framework)
```

Comment:
I thought these were going to hang off of `ProjectContext` directly. Then, only the `ProjectContext` itself really needs to be cached...